File: 238.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        vector<int> ans;
        int pr = 1, zer = 0;
        for(int j = 0; j < nums.size(); j++){
            if(nums[j] == 0){
                zer++;
                continue;
            }
            pr *= nums[j];
        }
        for(int i = 0; i < nums.size(); i++){
            if((zer > 0 && nums[i] != 0)){
                ans.push_back(0);
                continue;
            }
            if(nums[i] == 0){
                if(zer > 1)
                    ans.push_back(0);
                else 
                    ans.push_back(pr);
                continue;
            }
            ans.push_back(pr/nums[i]);
        }
        return ans;
    }
};

File: 238.go
package main

func productExceptSelf(nums []int) []int {
	prod := 1
	var ans []int
	z := 0
	for _, k := range nums {
		if k != 0 {
			prod *= k
		} else {
			z++
		}
	}
	for _, k := range nums {
		if k == 0 && z == 1 {
			ans = append(ans, prod)
		} else if z > 0 {
			ans = append(ans, 0)
		} else {
			ans = append(ans, prod/k)
		}
	}
	return ans
}
